src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -207,12 +207,32 @@ class LoweringToDatalogPass {
   static inline datalog::Predicate kDummyPredicate =
       datalog::Predicate("grounded_dummy", {"dummy_var"}, datalog::kPositive);
 
-  std::vector<datalog::Rule> QueriesToDLIR(
+  std::vector<datalog::RelationDeclaration> RelationDeclarationToDLIR(
+      const std::vector<datalog::RelationDeclaration>& relation_declaration);
+
+  std::vector<datalog::Rule> QueriesToDLIR(const std::vector<Query>& queries);
+
+  std::vector<datalog::RelationDeclaration> QueryRelationDeclarationToDLIR(
       const std::vector<Query>& queries);
 
   datalog::Program ProgToDLIR(const Program& program);
 
+  void update_can_say_depth(std::string_view predicate_name, uint64_t depth);
+
   uint64_t fresh_var_count_;
+  // This pass needs to generate relation declarations for predicates that
+  // appear in `AstCanSayFact`s which look like `A1 says A2 canSay ... An canSay
+  // foo(args)` and can in general be nested. The generated declarations look
+  // like: `says_cansay_..._cansay_foo(A1, A2, ..., An, args)`, so to generate
+  // these declarations, we need to know the number of "canSays" to include for
+  // each base predicate (e.g., foo). This pass does this by tracking the
+  // maximum depth of "canSay"s with which each predicate is used and storing
+  // this in the `cansay_depth` HashMap. When a predicate appears with depth
+  // `n`, it may also need to be used with depths (1..n), so we only maintain
+  // the maximum depth in which it appears and we generate declarations for
+  // (1..n). `cansay_depth` is only updated by `update_cansay_depth` which
+  // tracks this maximum.
+  absl::flat_hash_map<std::string_view, uint64_t> can_say_depth;
```

Comment:
C++ style: member variables hsould have a `_` at the end. Also `can_say_depths_` (plural) ?

`can_say_depth` ->`can_say_depths_`


Also, what does the key of this map represent?

Reply:
Key here is a predicate name. We check if there is any relation name that matches the predicate name and incrementally add new relation declarations based on the depth. Example:` P1 canSay P2 canSay grantAccess(x, y)`. Key here would be `grantAccess` with a depth value of 2